Extension/src/LanguageServer/copilotCompletionContextProvider.ts

```diff
@@ -0,0 +1,185 @@
+            telemetry.addError();
+            return [];
+        }
+    }
+
+    private async fetchTimeBudgetFactor(context: CompletionContext): Promise<number> {
+        const budgetFactor = context.activeExperiments.get("CppToolsCopilotTimeBudget");
+        return (budgetFactor as number) !== undefined ? budgetFactor as number : CopilotCompletionContextProvider.defaultTimeBudgetFactor;
+    }
+
+    public static async Create() {
+        const copilotCompletionProvider = new CopilotCompletionContextProvider();
+        await copilotCompletionProvider.registerCopilotContextProvider();
+        return copilotCompletionProvider;
+    }
+
+    public removeFile(fileUri: string): void {
+        this.completionContextCache.delete(fileUri);
+    }
+
+    public async resolve(context: CompletionContext, copilotAborts: vscode.CancellationToken): Promise<CodeSnippet[]> {
+        const startTime = performance.now();
+        const out: Logger = getOutputChannelLogger();
+        const timeBudgetFactor = await this.fetchTimeBudgetFactor(context);
+        const telemetry = new CopilotContextTelemetry();
+        let codeSnippets: CodeSnippet[] | undefined;
+        let codeSnippetsKind: SnippetsKind = SnippetsKind.Unknown;
+        try {
+            this.completionContextCancellation.cancel();
+            this.completionContextCancellation = new vscode.CancellationTokenSource();
+            const docUri = context.documentContext.uri;
+            const cachedValue: CodeSnippet[] | undefined = this.completionContextCache.get(docUri.toString());
+            const snippetsPromise = this.getCompletionContextWithCancellation(docUri,
+                context.documentContext.offset, startTime, out, telemetry.fork(), this.completionContextCancellation.token);
+            [codeSnippets, codeSnippetsKind] = await this.waitForCompletionWithTimeoutAndCancellation(
+                snippetsPromise, cachedValue, context.timeBudget * timeBudgetFactor, copilotAborts);
+            if (codeSnippetsKind === SnippetsKind.Cancelled) {
+                const duration: number = performance.now() - startTime;
+                out.appendLine(`Copilot: getCompletionContext(): cancelled, elapsed time (ms) : ${duration}`);
+                telemetry.addCancelled();
+                telemetry.addCancellationElapsed(duration);
+                throw new CancellationError();
+            }
+            telemetry.addSnippetCount(codeSnippets?.length);
+            return codeSnippets ?? [];
+        } catch (e: any) {
+            telemetry.addError();
+            throw e;
+        } finally {
+            telemetry.addKind(codeSnippetsKind.toString());
+            const duration: number = performance.now() - startTime;
+            if (codeSnippets === undefined) {
+                out.appendLine(`Copilot: getCompletionContext(): no snkppets provided (${codeSnippetsKind.toString()}), elapsed time (ms): ${duration}`);

```

Comment:
Can you round this duration to a whole number of ms? e.g. `83` instead of `83.27320000017062`.